module-services/service-cellular/Pdu.cpp

```diff
@@ -0,0 +1,972 @@
+        }
+
+        // http://www.openmobilealliance.org/release/Browser_Protocol_Stack/V2_1-20110315-A/OMA-WAP-TS-WSP-V1_0-20110315-A.pdf,
+        // p81
+        template <typename PageShiftPred, typename NamePred, typename ValuePred>
+        static bool parseHeader(Octets &octets, PageShiftPred pageShiftPred, NamePred namePred, ValuePred valuePred)
+        {
+            HeaderName name = std::uint8_t(0);
+            if (!parseGeneric<HeaderNamePolicy>(octets, name)) {
+                return false;
+            }
+
+            // Distinguish between page shift and short integer
+            std::uint32_t nameUint = 0;
+            std::visit(
+                [&](auto const &n) {
+                    using n_t = utils::remove_cref_t<decltype(n)>;
+                    if constexpr (std::is_same_v<n_t, std::uint8_t>) {
+                        nameUint = n;
+                    }
+                },
+                name);
+            if (nameUint > 0) {
+                if (nameUint < PageShift) {
+                    pageShiftPred(static_cast<std::uint8_t>(nameUint));
+                    return true;
+                }
+                else if (nameUint == PageShift) {
+                    std::uint8_t octet = 0;
+                    if (!octets.next(octet)) {
+                        return false;
+                    }
+                    pageShiftPred(octet);
+                    return true;
+                }
+                else if (nameUint > TextMax && nameUint <= 0xFF) {
+                    name = std::uint8_t(nameUint & 0x7F); // decode short integer
+                }
+                else if (nameUint > 0xFF) {
+                    return false; // This should not happen
+                }
+            }
+
+            if (!namePred(std::move(name))) {
+                int foo;
+                return parseGeneric<IgnoreHeaderValuePolicy>(octets, foo);
+            }
+
+            HeaderValue value = std::uint8_t(0);
+            if (!parseGeneric<DefaultPolicy>(octets, value)) {
+                return false;
+            }
+
+            valuePred(std::move(value));
+            return true;
+        }
+
+        template <std::size_t N>
+        static bool parseHeaders(Octets &octets,
+                                 std::array<KnownHeader, N> const &headerNames,
+                                 std::array<HeaderValue, N> &headerValues)
+        {
+            std::array<bool, N> headersFound;
+            headersFound.fill(false);
+            std::size_t foundCount = 0;
+
+            while (!octets.empty()) {
+                std::size_t nameId      = N;
+                HeaderValue headerValue = std::uint8_t(0);
+
+                auto pageShiftPred = [&](std::uint8_t page) {
+                    // TODO: Support page shifts
+                };
+                auto namePred = [&](auto &&name) {
+                    for (std::size_t i = 0; i < N; ++i) {
+                        if (headersFound[i]) {
```

Comment:
wouldn't `.at()` method be safer than brackets? I know that it probably won't go out of index, but you know :D

Reply:
In general yes, but here `N` is compile-time constant. I don't think that `at()` throwing an exception is needed here. Replacing the loop with recursive compile-time metafunction would be more complex than it's needed as well.

Reply:
Or we could use C++17 fold expressions instead of metafunction and check index at compile-time instead of throwing an exception, e.g.:
```
auto namePred = [&](auto &&name) {
    return any(
        [&](auto index) {
            static_assert(index < N);
            // Header was not found before and is required
            if (!headersFound[index] && isKnown(name, headerNames[index])) {
                headersFound[index] = true;
                ++foundCount;
                nameId = index;
                return true;
            }
            return false;
        },
        std::make_index_sequence<N>());
};
                
...

template <typename IndexPred, std::size_t... N>
static bool any(IndexPred indexPred, std::index_sequence<N...>)
{
    return (indexPred(std::integral_constant<std::size_t, N>()) || ...);
}
```

But I think this is not needed.

If you insist on checking `i` I'd rather use `assert(i < N)` instead of `at()` throwing and exception because if `i` was out of range it'd be programmer's fault, not an exceptional case. But I don't think this is needed either in this case.

Reply:
sure, you're right ;)